services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -269,6 +265,31 @@ public Response buildRefreshToken() {
         return Cors.add(request, Response.ok(res, MediaType.APPLICATION_JSON_TYPE)).auth().allowedOrigins(client).allowedMethods("POST").exposedHeaders(Cors.ACCESS_CONTROL_ALLOW_METHODS).build();
     }
 
+    private void updateClientSession(ClientSessionModel clientSession) {
+        if(clientSession == null) {
+            logger.error("client session is null");
+            return;
+        }
+        String adapterSessionId = formParams.getFirst(AdapterConstants.CLIENT_SESSION_STATE);
+        if (adapterSessionId != null) {
+            String adapterSessionHost = formParams.getFirst(AdapterConstants.CLIENT_SESSION_HOST);
+            logger.debugf("Adapter Session '%s' saved in ClientSession for client '%s'. Host is '%s'", adapterSessionId, client.getClientId(), adapterSessionHost);
+
+            event.detail(AdapterConstants.CLIENT_SESSION_STATE, adapterSessionId);
+            clientSession.setNote(AdapterConstants.CLIENT_SESSION_STATE, adapterSessionId);
+            event.detail(AdapterConstants.CLIENT_SESSION_HOST, adapterSessionHost);
+            clientSession.setNote(AdapterConstants.CLIENT_SESSION_HOST, adapterSessionHost);
+        }
+    }
+
+    private void updateClientSessions(List<ClientSessionModel> clientSessions) {
+        for (ClientSessionModel clientSession : clientSessions) {
+            if(client.getClientId().equals(client.getClientId())) {
```

Comment:
Thanks for the PR. Few comments:
- There is bug in this condition.  if(client.getClientId().equals(client.getClientId())) is always true. You may want to compare with clientId from the clientSession right?
- Is it possible to add the test for changing clientSessionState during refresh? We already have some tests as part of AdapterTest (and AdapterTestStrategy) class for classic logout, however I guess that you're not using keycloak adapters but you have some custom adapter code of your own?